doc/dev/design/NAT-address-discovery.rst

```diff
@@ -0,0 +1,143 @@
+NAT and the detector responds to the sender with a packet that contains the IP/port as seen by the detector.
+One example of this approach is the `STUN protocol <https://en.wikipedia.org/wiki/STUN>`_.
+
+One complication is that the local AS of the sender may be split into different subnets and that border routers
+are not all in the same subnet.
+Another complication is that some NATs may change their external port if a client endhosts connects to a new
+remote IP or port or uses a different local port.
+
+Therefore, it is desirable for the detector not only be in the same AS and same subnet as the BR, but ideally on the
+same server, listening on the same port.
+Similarly, it is desirable for the client to use the same local port for connection to the detector and for the
+actual data connection to the border router.
+
+Separately, we need specify a way how the client endhost can discover the "detector", i.e. its IP address and port.
+In case the detector shares and IP and possibly even port (or uses a fixed port) then this discovery is straight
+forward.
+If the port is flexible or the detector runs on a different IP then we need to find a different solution, probably
+as an extension to the discover/bootstrapping service.
+
+Proposal
+========
+[A precise statement of the proposed change.]
+
+Proposed change: Extend the border router (BR) to detect NATed addresses/ports and report them back to a client endhost.
+
+Ideally the solution would listen on the same port that would also be used for normal traffic forwarding.
+Alternatively, it may be feasible (depending on how sensitive the NAT is) to use a different (fixed) port to
+accept request for reporting addresses.
+
+The implementation on the protocol level could be done in several ways:
+
+1.  STUN: The BR needs to detect if an incoming packet is a STUN packet, and if it is, treat it is such and respond
+    to the STUN request.
+
+    -  Advantages: STUN is a well known and mature protocol. There are STUN libraries available in many programming languages.
+       It should be easy to implement this.
+
+2.  Extend SCMP with a new INFO message
+
+    -  Advantages: One less dependency on an external library and protocol
+    -  Disadvantages: More standardization effort? How do we solve authentication?
+
+3.  Extend SCMP with a new ERROR message: "invalid source address for first hop pkt", similar to error 33.
+    The router can verify that for first hop packets, the IP src address/port matches the SCION src address/port.
+    If not, it returns an error. The actual source address would need to be attached somewhere, unless we decide
+    to change the payload so it contains the IP header of the offending packet (and the IP header should contain the
+    NATed IP/port).
+
+    -  Advantages: One less dependency on an external library and protocol
+    -  Disadvantages: Conceptually a bit of a hack. More standardization effort? How do we solve authentication?
```

Comment:
Another clear disadvantage of this approach: The BR would need to check every outbound packet as part of the fast path.